### PR TITLE
Create and delete metallb configurations when running e2etests

### PR DESCRIPTION
**Pull request by sabinaaledort**
1. Adding before suite to configure MetalLB for testing by applying
   the MetalLB ConfigMap including both bgp and layer2 protocols.

2. One part of the configuration is runnig a BGP router in a container
   for all of the speakers to peer with.

3. Another part of the configuration is getting a range of 10 unused ip
   in the kind network subnet to create the layer2 address-pool.

4. Adding after suite to delete the MetalLB ConfigMap and kill the BGP
   router container.

Wasn't tested yet with dev-scpripts or OCP, might need some adjustments.

### CODE REVIEW

**Comment by russellb** [comment]
I'm not sure about running both `layer2` and `bgp` in the same configuration.  Once FRR is integrated, I expect we'll want to test `layer2` vs `bgp` using two different types of speaker deployments (with and without the FRR sidecar).

This also seems to introduce some duplication between the `e2etest` code and the `tasks.py` code for configuring MetalLB and FRR.  If there's good reason to be able to drive that from the test code, maybe it could be refactored so the dev-env setup could run just the configuration code?

**Comment by sabinaaledort** [comment]
> I'm not sure about running both `layer2` and `bgp` in the same configuration. Once FRR is integrated, I expect we'll want to test `layer2` vs `bgp` using two different types of speaker deployments (with and without the FRR sidecar).
> 
> This also seems to introduce some duplication between the `e2etest` code and the `tasks.py` code for configuring MetalLB and FRR. If there's good reason to be able to drive that from the test code, maybe it could be refactored so the dev-env setup could run just the configuration code?

i will remove the MetalLB configuration from `tasks.py`

**Comment by markdgray** [comment]
> I'm not sure about running both `layer2` and `bgp` in the same configuration. Once FRR is integrated, I expect we'll want to test `layer2` vs `bgp` using two different types of speaker deployments (with and without the FRR sidecar).
> 
> This also seems to introduce some duplication between the `e2etest` code and the `tasks.py` code for configuring MetalLB and FRR. If there's good reason to be able to drive that from the test code, maybe it could be refactored so the dev-env setup could run just the configuration code?

I think the `tasks.py` code is not very flexible to test all the test scenarios that we will need to run. For example, what if we want to test something that requires a configuration change to the FRR test instance and also MetalLB, we will need to redeploy both in a custom way. I guess, one option would be that we could make dev-env more configurable. However, I can imagine the list of command line options would start to get quite long. However, this would have the added advantage of being able to easily manually deploy the same configuration as what is being run in the test.

In this model, we would then run the e2etest go code which would exec.Command("inv dev-env --some-parameter --another-parameter") and then wait until the cluster was deployed before running the test for that configuration.

Alternatively, we could allow the e2etest go code, to run some docker commands in the FRR test instances in order to change the configuration. However, in this way, we will be unable to easily change the test topology (imagine a scenario in which we change the number of FRR test instances).

File: e2etest/config/frr/bgpd.conf [hunk 1]
```diff
@@ -0,0 +1,43 @@
+! -*- bgp -*-
```
Comment (markdgray) [inline e2etest/config/frr/bgpd.conf]: There are a few typos in the commit message.

Comment (markdgray) [reply e2etest/config/frr/bgpd.conf]: Also, could you update it to include the reason why you need to do this as I think that is the most important thing to get across.

File: dev-env/README.md [hunk 2]
```diff
@@ -10,19 +10,6 @@ of your git clone, which runs MetalLB in an unconfigured state:
 inv dev-env
 ```
```
Comment (markdgray) [inline dev-env/README.md:12]: We will still need to be able to run the development environment in order to manually test and try things out.

File: e2etest/e2etest_suite_test.go [hunk 3]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
 }
+
+var _ = ginkgo.BeforeSuite(func() {
+	cs, err := framework.LoadClientset()
+	framework.ExpectNoError(err)
+	_, err = cs.CoreV1().ConfigMaps(testNameSpace).Get(context.TODO(), "config", metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		err = configureMetalLB(cs)
+		framework.ExpectNoError(err)
+	} else if err == nil {
```
Comment (markdgray) [inline e2etest/e2etest_suite_test.go:94]: Maybe we should just fail if it has already been configured? That seems like an error condition.

File: e2etest/e2etest_suite_test.go [hunk 4]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
+	}
+})
+
+// Run a BGP router in a container for all of the speakers to peer with.
+func configureFRR(cs clientset.Interface) error {
+	// We need the ip of each node in the cluster to place them in the BGP router configuration file
+	// Each node will peer with this router.
+	nodes, err := cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+
+	// TODO: The IP address handling will need updates to add support for IPv6.
```
Comment (markdgray) [inline e2etest/e2etest_suite_test.go:116]: So this^ is required before merging?

File: e2etest/e2etest_suite_test.go [hunk 5]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
+	if err != nil {
+		return fmt.Errorf("Failed to create temp FRR config directory: %s", err)
+	}
+
+	err = CopyDir(consts.BGPTestFRRConfigDir, frrConfigTestDir)
+	if err != nil {
+		return fmt.Errorf("Failed to copy FRR config directory: %s", err)
+	}
+
+	bgpdConfigTestFile := fmt.Sprintf("%s/%s", frrConfigTestDir, consts.BGPDConfigFileName)
```
Comment (markdgray) [inline e2etest/e2etest_suite_test.go:141]: Should this be created in `consts.BGPTestFRRConfigDir`?

File: e2etest/e2etest_suite_test.go [hunk 6]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
+	}
+
+	err = ioutil.WriteFile(bgpdConfigTestFile, bgpdConfig, 0666)
+	if err != nil {
+		return fmt.Errorf("Failed to update FRR bgpd config file: %s", err)
+	}
+
+	// Run the BGP router in a container.
+	volume := fmt.Sprintf("%s:/etc/frr", frrConfigTestDir)
+	out, err := runCommand(false, "docker", "run", "-d", "--privileged", "--network", "kind", "--rm", "--name", consts.FRRContainerName, "--volume", volume, "frrouting/frr:v7.5.1")
```
Comment (markdgray) [inline e2etest/e2etest_suite_test.go:158]: We should add a const for the image name. e.g. frrouting/frr:v7.5.1. We will likely need to update this occasionally.

File: e2etest/e2etest_suite_test.go [hunk 7]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
+		}
+	}
+
+	return ipv4Range, ipv6Range, nil
+}
+
+// Configure MetalLB for testing.
+// Identify the unused network address range and use it in configmap.
+func configureMetalLB(cs clientset.Interface) error {
+	err := configureFRR(cs)
```
Comment (markdgray) [inline e2etest/e2etest_suite_test.go:227]: Will we need to reconfigure FRR? If we do, we may want to move `configureFRR()` out of this function.

File: e2etest/e2etest_suite_test.go [hunk 8]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
+	if err != nil {
+		return fmt.Errorf("Failed to get layer2 address pool range: %s", err)
+	}
+
+	peerAddress, err := runCommand(false, "docker", "inspect", "-f", "{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}", consts.FRRContainerName)
+	if err != nil {
+		return fmt.Errorf("Failed to get FRR ip address: %s", err)
+	}
+
+	// Apply the MetalLB ConfigMap.
```
Comment (markdgray) [inline e2etest/e2etest_suite_test.go:243]: It's probably not needed now, but we will need to configure MetalLB in different ways so I think the config map should be a parameter to this function.

File: e2etest/e2etest_suite_test.go [hunk 9]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
+		return fmt.Errorf("Failed to configure FRR: %s", err)
+	}
+
+	// TODO: should set IPv6 addresspool as well when adding support for IPv6.
+	ipv4Range, _, err := getL2AddressPoolRange()
+	if err != nil {
+		return fmt.Errorf("Failed to get layer2 address pool range: %s", err)
+	}
+
+	peerAddress, err := runCommand(false, "docker", "inspect", "-f", "{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}", consts.FRRContainerName)
```
Comment (markdgray) [inline e2etest/e2etest_suite_test.go:238]: Can you foresee a test case in which we have two different peers? i.e. two FRR Containers?

File: e2etest/e2etest_suite_test.go (hunk 8 above)
Comment (markdgray) [reply e2etest/e2etest_suite_test.go:243]: In general, I suspect that this function will need to be more configurable. It might be good to think about that now before we build on it.

File: e2etest/utils.go [hunk 10]
```diff
@@ -76,3 +84,150 @@ func tweakRCPort() func(rc *v1.ReplicationController) {
...
+func sortIPs(ips []string) []string {
+	sort.Slice(ips, func(i, j int) bool {
+		return compareIPs(ips[i], ips[j]) < 0
+	})
+
+	return ips
+}
+
+// Copy all the files in src directory to the dest directory.
+func CopyDir(src string, dest string) error {
```
Comment (markdgray) [inline e2etest/utils.go:191]: Could we just do an os call to `cp` or is the a go library that already does this?

File: e2etest/utils.go [hunk 11]
```diff
@@ -76,3 +84,150 @@ func tweakRCPort() func(rc *v1.ReplicationController) {
...
+	lastIP, err := incrementIPByValue(firstIP, network, rangeSize)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get ips range: %s", err)
+	}
+	ipsRange[1] = lastIP
+
+	return ipsRange, nil
+}
+
+// Compare two ip addresses.
```
Comment (markdgray) [inline e2etest/utils.go:149]: In what way, can you explain what comparison you are making?

File: e2etest/utils.go [hunk 12]
```diff
@@ -76,3 +84,150 @@ func tweakRCPort() func(rc *v1.ReplicationController) {
 	}
 	return nil
 }
+
+// Increment ip and test against CIDR after to avoid overflows.
```
Comment (markdgray) [inline e2etest/utils.go:88]: I didn't review the code below in detail but it seems like there may be go library functions that could help here?

File: e2etest/e2etest_suite_test.go (hunk 5 above)
Comment (sabinaaledort) [reply e2etest/e2etest_suite_test.go:141]: the frrConfigTestDir is a temp dir getting a random name starting with the prefix "frr-conf":
`frrConfigTestDir, err = ioutil.TempDir(".", "frr-conf")`

File: e2etest/e2etest_suite_test.go (hunk 4 above)
Comment (sabinaaledort) [reply e2etest/e2etest_suite_test.go:116]: im not sure about IPv6 support status. this wasn't support in the dev-env configuration either.

File: e2etest/utils.go (hunk 10 above)
Comment (sabinaaledort) [reply e2etest/utils.go:191]: i need to copy all the files from one directory to another so i use the read and write of "io/ioutil" go library. is there a better way?

File: e2etest/e2etest_suite_test.go [hunk 13]
```diff
@@ -70,3 +83,218 @@ func TestE2E(t *testing.T) {
...
+			if err != nil {
+				return nil, nil, fmt.Errorf("Failed to get ipv6 range: %s", err)
+			}
+		}
+	}
+
+	return ipv4Range, ipv6Range, nil
+}
+
+// Configure MetalLB for testing.
```
Comment (fedepaol) [inline e2etest/e2etest_suite_test.go:224]: This must be test-specific.
Different tests may need different configuration (and this is how we achieve coverage). The simplest example is having l2 tests / bgp tests, which may need different configurations, but also when we want to cover advanced configuraitons.

File: e2etest/utils.go (hunk 10 above)
Comment (fedepaol) [reply e2etest/utils.go:195]: +1

**Comment by fedepaol** [comment]
> I'm not sure about running both `layer2` and `bgp` in the same configuration. Once FRR is integrated, I expect we'll want to test `layer2` vs `bgp` using two different types of speaker deployments (with and without the FRR sidecar).

Agree on the configuration side. The configuration must be driven by the tests and (apart from willingly wanting to test the corner case of both l2 / bpg configured together) each test (or test class) must apply (and clean) its own configuration.

Re: the two types of deployments, I expect them to be driven from outside and be two different CI lanes, one with the current bgp implementation and the other with FRR. The test suite must be the same and not know about the implementation (unless we start introducing frr-only features)

File: e2etest/utils.go [hunk 14]
```diff
@@ -1,11 +1,16 @@
 package e2e
 
 import (
+	"context"
 	"fmt"
 	"os/exec"
 
+	pkgerrors "github.com/pkg/errors"
```
Comment (fedepaol) [inline e2etest/utils.go:8]: Why you need to alias it? `errors` is generally intended to be this one, I'd alias anything else conflicts with this.

File: e2etest/utils.go [hunk 15]
```diff
@@ -76,3 +136,31 @@ func tweakRCPort() func(rc *v1.ReplicationController) {
 	}
 	return nil
 }
+
+func updateConfigMapObjs(cs clientset.Interface, dataModifier func(m *configFile)) error {
```
Comment (fedepaol) [inline e2etest/utils.go:140]: I think the whole config parsing stuff deserves a separate package.
We should aim to keep the utils.go file slim / remove it, not adding new stuff to it.
https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common

File: e2etest/utils.go [hunk 16]
```diff
@@ -76,3 +136,31 @@ func tweakRCPort() func(rc *v1.ReplicationController) {
...
+		cm.Data = make(map[string]string)
+	}
+
+	data := configFile{}
+	err = yaml.Unmarshal([]byte(cm.Data["config"]), &data)
+	framework.ExpectNoError(err)
+
+	dataModifier(&data)
+	resData, err := yaml.Marshal(data)
+	framework.ExpectNoError(err)
```
Comment (fedepaol) [inline e2etest/utils.go:156]: We are mixing framework.ExpectNoError with returning the error.
Just pick one.

File: e2etest/utils.go [hunk 17]
```diff
@@ -76,3 +136,31 @@ func tweakRCPort() func(rc *v1.ReplicationController) {
...
 }
+
+func updateConfigMapObjs(cs clientset.Interface, dataModifier func(m *configFile)) error {
+	cm, err := cs.CoreV1().ConfigMaps(testNameSpace).Get(context.TODO(), "config", metav1.GetOptions{})
+	if err != nil {
+		return pkgerrors.Wrapf(err, "Failed to get MetalLB ConfigMap")
+	}
+
+	if len(cm.Data) == 0 {
+		cm.Data = make(map[string]string)
```
Comment (fedepaol) [inline e2etest/utils.go:147]: Do you need to do this? Here we are always overriding the whole config right?

Comment (fedepaol) [reply e2etest/utils.go:147]: Also, wouldn't everything be simpler if we always delete and create the configmap with the new config? Am I missing something?

File: e2etest/consts/const.go [hunk 18]
```diff
@@ -0,0 +1,12 @@
+package consts
+
+const (
+	// BGPTestFRRConfigDir contains the name of the BGP configuration directory
+	BGPTestFRRConfigDir = "config/frr"
```
Comment (fedepaol) [inline e2etest/consts/const.go:5]: Please move the constants used in a single package to that package. We need a shared package only if multiple packages access the same constant.

File: e2etest/bgp_test.go [hunk 19]
```diff
@@ -80,8 +131,117 @@ var _ = ginkgo.Describe("BGP", func() {
...
+
+		err = wgetRetry(useDocker, frrContainerTestName, address)
 		framework.ExpectNoError(err)
 	})
 
 })
+
+// Run a BGP router in a container.
+func startFRRContainer(containerName string, testDirName string) (string, error) {
+	srcFiles := fmt.Sprintf("%s/.", consts.BGPTestFRRConfigDir)
```
[the remark this answers is not in the document]
Comment (sabinaaledort) [reply e2etest/bgp_test.go:143]: just tried that, apparently it removes dots. filepath.Join() is calling filepath.Clean() which eliminates each dot path name element (the current directory).

File: e2etest/bgp_test.go [hunk 20]
```diff
@@ -21,38 +21,89 @@ package e2e
...
+		err := bgpCS.CoreV1().ConfigMaps(testNameSpace).Delete(context.TODO(), "config", metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+
 		if ginkgo.CurrentGinkgoTestDescription().Failed {
 			DescribeSvc(f.Namespace.Name)
 		}
 	})
 
 	ginkgo.It("should work for type=Loadbalancer", func() {
+		frrContainerTestName := "frr-test"
```
[the remark this answers is not in the document]
Comment (sabinaaledort) [reply e2etest/bgp_test.go:79]: i think the name should be set in the test as we might have a test with 2 frr containers.

File: e2etest/bgp_test.go [hunk 21]
```diff
@@ -80,8 +131,117 @@ var _ = ginkgo.Describe("BGP", func() {
...
+		return "", errors.Wrapf(err, "Failed to get FRR ip address")
+	}
+
+	containerIP = strings.Split(containerIP, "\n")[0]
+
+	return containerIP, nil
+}
+
+// Delete the BGP router configuration.
+func deleteFRRConfiguration(containerName string, testDirName string) error {
```
[the remark this answers is not in the document]
Comment (sabinaaledort) [reply e2etest/bgp_test.go]: what config reloading mechanism do you mean?

File: tasks.py [hunk 22]
```diff
@@ -385,6 +385,38 @@ def bgp_dev_env():
     # Apply the MetalLB ConfigMap
     run("kubectl apply -f %s/config.yaml" % dev_env_dir)
 
+def get_service_range(networkVersion=None):
```
[the remark this answers is not in the document]
Comment (sabinaaledort) [reply tasks.py:388]: do you mean use ipv4 and ipv6 instead of 4 and 6?

File: e2etest/bgp_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (fedepaol) [reply e2etest/bgp_test.go:64]: If / when we have a test with two containers, we may have a set of tests depending on them.
Creating the external container (or containers) only once, will make the test execution quicker.

File: e2etest/config/frr/zebra.conf [hunk 23]
```diff
@@ -0,0 +1,3 @@
+hostname Router
```
Comment (fedepaol) [inline e2etest/config/frr/zebra.conf:1]: Do we need this at all?

File: e2etest/bgp_container.go [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+	"strings"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+const (
+	// BGPTestFRRConfigDir contains the name of the BGP configuration directory.
+	BGPTestFRRConfigDir = "config/frr"
```
Comment (fedepaol) [inline e2etest/bgp_container.go:36]: If the consts are package local, no need to export them

File: e2etest/bgp_container.go [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+		}
+	}
+
+	if len(nodesIPs) < 3 {
+		return "", fmt.Errorf("Expected 3 nodes, got %d", len(nodesIPs))
+	}
+
+	bgpPeersInfo := "router bgp 64512\n"
+	for _, ip := range nodesIPs {
+		bgpPeersInfo = fmt.Sprintf("%s neighbor %s remote-as 64512\n neighbor %s next-hop-self\n", bgpPeersInfo, ip, ip)
```
Comment (fedepaol) [inline e2etest/bgp_container.go:154]: Is the `remote-as` equal the local router id?

File: e2etest/bgp_container.go [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+	nodesIPs := make([]string, 0, len(nodes.Items))
+	for _, node := range nodes.Items {
+		for i := range node.Status.Addresses {
+			if node.Status.Addresses[i].Type == "InternalIP" {
+				nodesIPs = append(nodesIPs, node.Status.Addresses[i].Address)
+			}
+		}
+	}
+
+	if len(nodesIPs) < 3 {
```
Comment (fedepaol) [inline e2etest/bgp_container.go:148]: I'd remove this constraint, at this point we don't really care about the number of nodes.

File: e2etest/bgp_container.go [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+		return "", fmt.Errorf("Expected 3 nodes, got %d", len(nodesIPs))
+	}
+
+	bgpPeersInfo := "router bgp 64512\n"
```
Comment (fedepaol) [inline e2etest/bgp_container.go:152]: How about using a template here too (with a local template, no need to have an external file)?

File: e2etest/bgp_test.go [hunk 28]
```diff
@@ -44,12 +47,56 @@ var _ = ginkgo.Describe("BGP", func() {
 	})
 
 	ginkgo.AfterEach(func() {
+		err := cs.CoreV1().ConfigMaps(testNameSpace).Delete(context.TODO(), "config", metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+
+		Eventually(func() bool {
+			_, err := cs.CoreV1().ConfigMaps(testNameSpace).Get(context.TODO(), "config", metav1.GetOptions{})
+			return apierrr.IsNotFound(err)
+		}, 1*time.Minute, 5*time.Second).Should(BeTrue(), "Failed to delete MetalLB ConfigMap")
```
Comment (fedepaol) [inline e2etest/bgp_test.go:56]: Let's hope it's quicker. Checking every minute will make this at least one minute slow.
Move to 1 sec

File: e2etest/bgp_container.go [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// https://github.com/kubernetes/kubernetes/blob/92aff21558831b829fbc8cbca3d52edc80c01aa3/test/e2e/network/loadbalancer.go#L878
+
+package e2e
```
Comment (fedepaol) [inline e2etest/bgp_container.go:19]: I think this deserves to be in its own package.

I'd choose `frr` for the container operations, `frrconfig` or just `config` under frr for the config operations

File: e2etest/bgp_container.go [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+
+import (
+	"context"
+	"fmt"
+	"html/template"
```
Comment (fedepaol) [inline e2etest/bgp_container.go:24]: Use text/template

File: e2etest/bgp_container.go [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "Failed to change %s container volume permissions. %s", containerName, string(out))
+	}
+
+	return nil
+}
+
+// Delete the BGP router configuration.
+func stopFRRContainer(containerName string, testDirName string) error {
+	err := updateFRRContainerVolumePermissions(containerName)
```
Comment (fedepaol) [inline e2etest/bgp_container.go:111]: Calling this right after the creation would make it possible to edit the file as part of the tests, right?

File: e2etest/bgp_container.go [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+	BGPDConfigFileName = "bgpd.conf"
+	// FRRRoutingImage contains the name and tag of the FRR routing image.
+	FRRRoutingImage = "frrouting/frr:v7.5.1"
+	// FRRContainerMountDestination contains the mount destination path.
+	FRRContainerMountDestination = "/etc/frr"
+)
+
+// Run a BGP router in a container.
+// Returns the IPv4 and IPv6 addresses of the container.
+func startFRRContainer(cs clientset.Interface, containerName string, testDirName string, bgpConfig string) (string, string, error) {
```
Comment (fedepaol) [inline e2etest/bgp_container.go:47]: I would let this just start the container with an empty config file, and implement another method `reloadFRRConfig` that basically leverages what @oribon used in https://github.com/metallb/metallb/pull/935 .
This will make handling external containers more flexible, and will also make it easier to use the same set of tests when we are running frr on localhost.

File: e2etest/bgp_test.go (hunk 28 above)
Comment (sabinaaledort) [reply e2etest/bgp_test.go:56]: it is checking every 5 seconds, for 1 minute

Comment (fedepaol) [reply e2etest/bgp_test.go:56]: Ooops :-) Can we move it to 1 second?

File: e2etest/bgp_test.go [hunk 33]
```diff
@@ -49,12 +54,71 @@ var _ = ginkgo.Describe("BGP", func() {
...
+				},
+			},
+		}
+		err = updateConfigMap(cs, configData)
+		framework.ExpectNoError(err)
+
+		bgpConfig, err := frrconfig.BGPPeersForAllNodes(cs)
+		framework.ExpectNoError(err)
+
+		err = frr.UpdateBGPConfigFile(frrContainer, frrTestConfigDir, bgpConfig)
```
[the remark this answers is not in the document]
Comment (sabinaaledort) [reply e2etest/bgp_test.go:103]: i thought we might want to reload other files as well so i separated these two.

[the remark this answers is not in the document]
Comment (fedepaol) [reply e2etest/bgp_test.go:103]: That won't prevent us to have the reload function called by the update one. I can't think about a case where we update the config but we don't want to reload it

[the remark this answers is not in the document]
Comment (fedepaol) [reply e2etest/bgp_test.go:103]: so is there any reason why update can't also reload using the other function inside?

[the remark this answers is not in the document]
Comment (sabinaaledort) [reply e2etest/bgp_test.go:103]: sounds good :+1:

**Comment by fedepaol** [comment]
This will serve as a base for extending the coverage by changing configurations inside the tests.

LGTM
cc @russellb
